### PR TITLE
AAG-5 ViewLayer

### PR DESCRIPTION
**Pull request by MauBocanegraTorres**
[MVI ViewLayer](https://distillery.atlassian.net/browse/AAG-5)

Added an example of the implementation to call the flow to communicate to the next layer
The classes inside `mvi_example` can be removed if doesn't apply to this PR

- **Tests pending**

**Checklist:**
- [ ] This PR follows [guidelines](https://github.com/DistilleryTech/android-arch-blueprints/wiki/Pull-Request-Process)
- [ ] I have performed a self-review of my own code
- [ ] I documented all new code and updated the external docs if necessary
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by cnavarrete101** [comment]
Change target branch

**Comment by MauBocanegraTorres** [comment]
> Change target branch

Thats why all the files 
Let me restore everything then

File: mvi/src/main/java/models/StateResource.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package models
+
+data class StateMessage(val response: Response)
+
+data class Response(
```
Comment (lucaslabs) [inline mvi/src/main/java/models/StateResource.kt:5]: Can you please put first the sealed/parent class (MessageType) then its subclasses, for legibility purpose.

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/intents/TodoListInteractionManager.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package com.distillery.android.blueprints.mvi.todo.intents
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import com.distillery.android.domain.models.ToDoModel
+
+class TodoListInteractionManager {
```
Comment (lucaslabs) [inline mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/intents/TodoListInteractionManager.kt:7]: This is an Intent, Interactor or Interaction? 🤔

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/models/states/TodoState.kt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package com.distillery.android.blueprints.mvi.todo.models.states
+
```
Comment (cnavarrete101) [inline mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/models/states/TodoState.kt:2]: I think a comment here could help to understand that this state belongs to the Adapter that will be used

File: mvi/src/main/java/models/states/DataState.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import models.StateMessage
+
+data class DataState<T>(
+    var stateMessage: StateMessage? = null,
+    var data: T? = null,
+    var stateEvent: StateEvent? = null
+) {
+    companion object {
+        fun <T> data(
+            response: Response?,
```
Comment (cnavarrete101) [inline mvi/src/main/java/models/states/DataState.kt:13]: I think Response is unnecessary, thinking in regard to this TODO app I believe data could be inside the stateEvent.

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/viewmodels/TodoListViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import models.Response
+import models.StateMessage
+import models.states.DataState
+import models.states.ViewState
+import models.states.StateEvent
+import viewmodel.BaseViewModel
+
+@FlowPreview
+@ExperimentalCoroutinesApi
+class TodoListViewModel : BaseViewModel<TodoListViewState>() {
```
Comment (cnavarrete101) [inline mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/viewmodels/TodoListViewModel.kt:20]: Probably since it is not possible to see the whole path of the file could be helpfull to add a comment mentioning this is an implementation, as same as other classes with the prefix ToDo

**Comment by cnavarrete101** [comment]
@distillery-alexander-ageychenko @gorrotowi what do you guys think? regarding DataState and Response classes

File: mvi/src/main/java/models/StateResource.kt (hunk 1 above)
Comment (distillery-alexander-ageychenko) [reply mvi/src/main/java/models/StateResource.kt:5]: @lucaslabs could you rephrase a bit? Not quite clear what is not right from your point of view, as for me this class looks totally fine

File: mvi/src/main/java/models/states/DataState.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package models.states
+
+import models.Response
+import models.StateMessage
+
+data class DataState<T>(
```
Comment (distillery-alexander-ageychenko) [inline mvi/src/main/java/models/states/DataState.kt:6]: Would be great to have descriptive docs for each class so it will be clear why do we need it and what's its purpose

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/viewmodels/TodoListViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    private val dataChannel: kotlinx.coroutines.flow.Flow<DataState<ViewState>> = flow {}
+    val todoListInteractionManager = TodoListInteractionManager()
+
+    override fun initNewViewState(): com.distillery.android.blueprints.mvi.todo.models.states.TodoListViewState {
+        return com.distillery.android.blueprints.mvi.todo.models.states.TodoListViewState()
+    }
+
+    // each case downloads from repository and returns a [DataState] object
+    // here we are mocking the flows and this SHOULD be updated to receive only DataState
+    override fun setStateEvent(
```
Comment (distillery-alexander-ageychenko) [inline mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/viewmodels/TodoListViewModel.kt:31]: if I'm getting it right, this `StateEvent` is something that view will generate and pass to the view model, which wil decide what to do based on it's type? If so then it does look like `intent` from the MVI. So we have an Intent to do something and we pass that intent to the view model, it passed intent to some business logic where we take that intent with current state and produce an updated state (reducer logic)

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/intents/TodoListInteractionManager.kt (hunk 2 above)
Comment (distillery-alexander-ageychenko) [inline mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/intents/TodoListInteractionManager.kt:7]: Minor: suffix "Manager" doesn't add any meaning to the class name, what does it manage? From the code I'd say it looks like "ToDoListRepository" as it provides collection like access to the data

File: mvi/src/main/java/models/StateResource.kt (hunk 1 above)
Comment (cnavarrete101) [reply mvi/src/main/java/models/StateResource.kt:5]: what @lucaslabs meant is that the sealed class should be at the top of the file and not at the bottom

Comment (distillery-alexander-ageychenko) [reply mvi/src/main/java/models/StateResource.kt:5]: @cnavarrete101 why though? Haven't heard about such practice, at top completely different classes that are not subclasses of the sealed class, it's subclasses are inside of it.

File: mvi/src/main/java/models/states/DataState.kt [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    var stateMessage: StateMessage? = null,
+    var data: T? = null,
+    var stateEvent: StateEvent? = null
```
Comment (distillery-alexander-ageychenko) [inline mvi/src/main/java/models/states/DataState.kt:9]: as I understand data is the core thing that we need, state message is some kind of description of current state, how we can use it? and what is the purpose of state event?

File: mvi/src/main/java/models/states/DataState.kt [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    companion object {
+        fun <T> data(
+            response: Response?,
+            data: T? = null,
+            stateEvent: StateEvent?
+        ): DataState<T> {
+            return DataState(
+                stateMessage = response?.let {
+                    StateMessage(it)
+                },
```
Comment (distillery-alexander-ageychenko) [inline mvi/src/main/java/models/states/DataState.kt:20]: not quiet clear why response becomes state message, we can introduce some kind of extension fun with meaningful name in order to make it better, but some description of mapping logic would be great as well

File: mvi/src/main/java/models/states/DataState.kt [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+import models.Response
+import models.StateMessage
+
+data class DataState<T>(
+    var stateMessage: StateMessage? = null,
+    var data: T? = null,
+    var stateEvent: StateEvent? = null
+) {
+    companion object {
+        fun <T> data(
```
Comment (distillery-alexander-ageychenko) [inline mvi/src/main/java/models/states/DataState.kt:12]: @MauBocanegraTorres  DataState.data(...) would look a bit weird to me if seen in code, I think we can improve naming by mentioning goal of the method, as I see we create a DataState object, so `fun create(): DataState` looks reasonable, what do you think?

File: mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/models/states/TodoListStateEvent.kt [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import models.states.StateEvent
+
+sealed class TodoListStateEvent : StateEvent {
+
+    class InsertNewTodoEvent(
+        val description: String
+    ) : TodoListStateEvent() {
+
+        override fun eventName(): String {
+            return "InsertNewTodoEvent"
```
Comment (distillery-alexander-ageychenko) [inline mvi_example/src/main/java/com/distillery/android/blueprints/mvi/todo/models/states/TodoListStateEvent.kt:12]: what is the purpose of string with name? It looks to me that we can use class name instead, if we don't plan to use some other string instead

File: mvi/src/main/java/models/StateResource.kt (hunk 1 above)
Comment (cnavarrete101) [reply mvi/src/main/java/models/StateResource.kt:5]: I don't know, it was Lucas comment :P

**Comment by distillery-alexander-ageychenko** [comment]
PR loos stale and probably obsolete, will close for now
